libs/utils/src/pg_checksum_page.rs

```diff
@@ -0,0 +1,120 @@
+///
+///  Port of Postgres pg_checksum_page
+///
+
+const BLCKSZ: usize = 8192;
+const N_SUMS: usize = 32;
+/* prime multiplier of FNV-1a hash */
+const FNV_PRIME: u32 = 16777619;
+
+/*
+ * Base offsets to initialize each of the parallel FNV hashes into a
+ * different initial state.
+ */
+const CHECKSUM_BASE_OFFSETS: [u32; N_SUMS] = [
+    0x5B1F36E9, 0xB8525960, 0x02AB50AA, 0x1DE66D2A, 0x79FF467A, 0x9BB9F8A3, 0x217E7CD2, 0x83E13D2C,
+    0xF8D4474F, 0xE39EB970, 0x42C6AE16, 0x993216FA, 0x7B093B5D, 0x98DAFF3C, 0xF718902A, 0x0B1C9CDB,
+    0xE58F764B, 0x187636BC, 0x5D7B3BB1, 0xE73DE7DE, 0x92BEC979, 0xCCA6C0B2, 0x304A0979, 0x85AA43D4,
+    0x783125BB, 0x6CA8EAA2, 0xE407EAC6, 0x4B5CFC3E, 0x9FBF8C76, 0x15CA20BE, 0xF2CA9FD3, 0x959BD756,
+];
+
+/*
+ * Calculate one round of the checksum.
+ */
+fn checksum_comp(checksum: u32, value: u32) -> u32 {
+    let tmp = checksum ^ value;
+    tmp.wrapping_mul(FNV_PRIME) ^ (tmp >> 17)
+}
+
+/*
+ * Compute the checksum for a Postgres page.
+ *
+ * The page must be adequately aligned (at least on a 4-byte boundary).
+ * Beware also that the checksum field of the page is transiently zeroed.
+ *
+ * The checksum includes the block number (to detect the case where a page is
+ * somehow moved to a different location), the page header (excluding the
+ * checksum itself), and the page data.
+ *
+ * As in C implementation in Postgres, the checksum attribute on the page is
+ * excluded from the calculation and preserved.
+ */
+pub fn pg_checksum_page(data: &[u8], blkno: u32) -> u16 {
+    let page = unsafe { std::mem::transmute::<&[u8], &[u32]>(data) };
```

Comment:
One possible way to fix current code is to use [chunks_exact(4)](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact) and mapping to [u32::from_ne_bytes](https://doc.rust-lang.org/std/primitive.u32.html#method.from_ne_bytes). And check that chunks iterator remainder is empty or assert on buffer size before creating an iterator

Reply:
Hm, the fact that Postgres hashing algorithm is based on FNV doesn't mean that it's identical. The comment in Postgres says:
```clang
 * PostgreSQL doesn't use FNV-1a hash directly because it has bad mixing of
 * high bits - high order bits in input data only affect high order bits in
 * output data. To resolve this we xor in the value prior to multiplication
 * shifted right by 17 bits. 
 ```
 so I'm afraid that it will just produce different hashes and Postgres won't accept these checksums. We can try this fnv crate, I don't really think that it work out, do you?
 
 > This wont work for all possible inputs.
 
 Yes, because the page size is fixed to the 8 KB, we can add size check though, just to be on the safe side, but it won't change anything really, I suppose.

Reply:
> One possible way to fix current code is to use [chunks_exact(4)](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact) and mapping to [u32::from_ne_bytes](https://doc.rust-lang.org/std/primitive.u32.html#method.from_ne_bytes). And check that chunks iterator remainder is empty or assert on buffer size before creating an iterator

`u32::from_ne_bytes` should work, but won't it be slower? It should involve copying the whole page chunk by chunk each time, which isn't a desired behavior as checksum calculation may happen quite often. Or I miss something and it's optimized / or we can optimize its usage?